### DISCUSSION TITLE
Where `llama.cpp` releases `ggml_cgraph * gf = llama_build_graph(lctx, ubatch, false);`

### DISCUSSION THREAD

**Discussion by Hlinbit** [ggml-org/llama.cpp#10482] [Q&A]
Recently I have been reading and debugging the code of `llama.cpp`. 

When I read `static int llama_decode_internal`, I found that the calculation graph pointer (ordinary pointer, non-smart pointer):  `ggml_cgraph * gf = llama_build_graph(lctx, ubatch, false);`,  built in the function was not released. 

Wouldn't this cause a memory leak? 

Since I just started studying these codes, I may not know enough about them.

**Comment by Hlinbit** [comment]
After reading the function `static void ggml_backend_sched_split_graph(ggml_backend_sched_t sched, struct ggml_cgraph * graph)`, I have some new understanding.

The splitting process is also a storing process. The function splits the graph and stores the subgraph into `sched->splits`. After splitting, the process can find the computing graph from the local variable `gf` in `llama_decode_internal` and from the `sched->splits`. And the process will release the computing graph through `ggml_backend_sched_ptr sched` in `llama_context`.

**Comment by Hlinbit** [comment]
Brothers, is my understanding correct?

**Comment by slaren** [comment]
Like the tensors, the graph is allocated from the memory in the `ggml_context`.

**Reply by Hlinbit** [reply]
It seems like almost all resources are allocated in the memory pool `ggml_context`. `ggml_context` supports safe memory management for llama.cpp.
